File: languages/csharp/Bitwarden.Sdk/ProjectsClient.cs
namespace Bitwarden.Sdk;

public class ProjectsClient
{
    private readonly CommandRunner _commandRunner;

    internal ProjectsClient(CommandRunner commandRunner)
    {
        _commandRunner = commandRunner;
    }

    public async Task<ProjectResponse> GetAsync(Guid id, CancellationToken cancellationToken = default)
    {
        var command = new Command { Projects = new ProjectsCommand { Get = new ProjectGetRequest { Id = id } } };
        var result = await _commandRunner.RunCommandAsync<ResponseForProjectResponse>(command, cancellationToken);

        if (result is { Success: true })
        {
            return result.Data;
        }

        throw new BitwardenException(result != null ? result.ErrorMessage : "Project not found");
    }

    public async Task<ProjectResponse> CreateAsync(Guid organizationId, string name, CancellationToken cancellationToken = default)
    {
        var command = new Command
        {
            Projects = new ProjectsCommand
            {
                Create = new ProjectCreateRequest { OrganizationId = organizationId, Name = name }
            }
        };
        var result = await _commandRunner.RunCommandAsync<ResponseForProjectResponse>(command, cancellationToken);

        if (result is { Success: true })
        {
            return result.Data;
        }

        throw new BitwardenException(result != null ? result.ErrorMessage : "Project create failed");
    }

    public async Task<ProjectResponse> UpdateAsync(Guid organizationId, Guid id, string name, CancellationToken cancellationToken = default)
    {
        var command = new Command
        {
            Projects = new ProjectsCommand
            {
                Update = new ProjectPutRequest { Id = id, OrganizationId = organizationId, Name = name }
            }
        };
        var result = await _commandRunner.RunCommandAsync<ResponseForProjectResponse>(command, cancellationToken);

        if (result is { Success: true })
        {
            return result.Data;
        }

        throw new BitwardenException(result != null ? result.ErrorMessage : "Project update failed");
    }

    public async Task<ProjectsDeleteResponse> DeleteAsync(Guid[] ids, CancellationToken cancellationToken = default)
    {
        var command = new Command
        {
            Projects = new ProjectsCommand { Delete = new ProjectsDeleteRequest { Ids = ids } }
        };
        var result = await _commandRunner.RunCommandAsync<ResponseForProjectsDeleteResponse>(command, cancellationToken);

        if (result is { Success: true })
        {
            return result.Data;
        }

        throw new BitwardenException(result != null ? result.ErrorMessage : "Project delete failed");
    }

    public async Task<ProjectsResponse> ListAsync(Guid organizationId, CancellationToken cancellationToken = default)
    {
        var command = new Command
        {
            Projects = new ProjectsCommand { List = new ProjectsListRequest { OrganizationId = organizationId } }
        };
        var result = await _commandRunner.RunCommandAsync<ResponseForProjectsResponse>(command, cancellationToken);

        if (result is { Success: true })
        {
            return result.Data;
        }

        throw new BitwardenException(result != null ? result.ErrorMessage : "No projects for given organization");
    }
}

File: languages/java/src/main/java/com/bitwarden/sdk/ProjectsClient.java
package com.bitwarden.sdk;

import com.bitwarden.sdk.schema.*;

import java.util.UUID;

public class ProjectsClient {

    private final CommandRunner commandRunner;

    ProjectsClient(CommandRunner commandRunner) {
        this.commandRunner = commandRunner;
    }

    public ProjectResponse get(UUID id) {
        Command command = new Command();
        ProjectsCommand projectsCommand = new ProjectsCommand();
        ProjectGetRequest projectGetRequest = new ProjectGetRequest();
        projectGetRequest.setID(id);
        projectsCommand.setGet(projectGetRequest);
        command.setProjects(projectsCommand);

        ResponseForProjectResponse response = commandRunner.runCommand(command,
            BitwardenClient.throwingFunctionWrapper(Converter::ResponseForProjectResponseFromJsonString));

        if (response == null || !response.getSuccess()) {
            throw new BitwardenClientException(response != null ? response.getErrorMessage() : "Project not found");
        }

        return response.getData();
    }

    public ProjectResponse create(UUID organizationId, String name) {
        Command command = new Command();
        ProjectsCommand projectsCommand = new ProjectsCommand();
        ProjectCreateRequest projectCreateRequest = new ProjectCreateRequest();
        projectCreateRequest.setOrganizationID(organizationId);
        projectCreateRequest.setName(name);
        projectsCommand.setCreate(projectCreateRequest);
        command.setProjects(projectsCommand);

        ResponseForProjectResponse response = commandRunner.runCommand(command,
            BitwardenClient.throwingFunctionWrapper(Converter::ResponseForProjectResponseFromJsonString));

        if (response == null || !response.getSuccess()) {
            throw new BitwardenClientException(response != null ? response.getErrorMessage() : "Project create failed");
        }

        return response.getData();
    }

    public ProjectResponse update(UUID organizationId, UUID id, String name) {
        Command command = new Command();
        ProjectsCommand projectsCommand = new ProjectsCommand();
        ProjectPutRequest projectPutRequest = new ProjectPutRequest();
        projectPutRequest.setID(id);
        projectPutRequest.setOrganizationID(organizationId);
        projectPutRequest.setName(name);
        projectsCommand.setUpdate(projectPutRequest);
        command.setProjects(projectsCommand);

        ResponseForProjectResponse response = commandRunner.runCommand(command,
            BitwardenClient.throwingFunctionWrapper(Converter::ResponseForProjectResponseFromJsonString));

        if (response == null || !response.getSuccess()) {
            throw new BitwardenClientException(response != null ? response.getErrorMessage() : "Project update failed");
        }

        return response.getData();
    }

    public ProjectsDeleteResponse delete(UUID[] ids) {
        Command command = new Command();
        ProjectsCommand projectsCommand = new ProjectsCommand();
        ProjectsDeleteRequest projectsDeleteRequest = new ProjectsDeleteRequest();
        projectsDeleteRequest.setIDS(ids);
        projectsCommand.setDelete(projectsDeleteRequest);
        command.setProjects(projectsCommand);

        ResponseForProjectsDeleteResponse response = commandRunner.runCommand(command,
            BitwardenClient.throwingFunctionWrapper(Converter::ResponseForProjectsDeleteResponseFromJsonString));

        if (response == null || !response.getSuccess()) {
            throw new BitwardenClientException(response != null ?
                response.getErrorMessage() : "Projects update failed");
        }

        return response.getData();
    }

    public ProjectsResponse list(UUID organizationId) {
        Command command = new Command();
        ProjectsCommand projectsCommand = new ProjectsCommand();
        ProjectsListRequest projectsListRequest = new ProjectsListRequest();
        projectsListRequest.setOrganizationID(organizationId);
        projectsCommand.setList(projectsListRequest);
        command.setProjects(projectsCommand);

        ResponseForProjectsResponse response = commandRunner.runCommand(command,
            BitwardenClient.throwingFunctionWrapper(Converter::ResponseForProjectsResponseFromJsonString));

        if (response == null || !response.getSuccess()) {
            throw new BitwardenClientException(response != null ?
                response.getErrorMessage() : "No projects for given organization");
        }

        return response.getData();
    }
}
